doc/decisions/1_in_discussion/change_tracking.md

```diff
@@ -0,0 +1,181 @@
+  It may, however, be debatable whatever 'minimal' means in this context.
+- Adding fields to the structs causes a slight memory overhead, even with tracking turned off.
+  While negligible for `KeySet` due to the low amount of keysets in typical applications, it may be noticable for `Key`.
+  On a 64-bit system we'd add 16 bytes to it.
+  8 bytes for the pointer to the original value, 8 bytes for the size of the original value.
+  To put this in perspective, the current size of the `Key` struct is 64 bytes, so we'd add 25% overhead to an empty key.
+  However, this percentage will be much lower in a real-world application, as the usefulness of an empty key is very low.
+- Another downside here is that it is not so easy to determine what the "original" value is.
+  Some part of `libelektra-kdb` would need to mark the keys as original (after transformations etc.)
+
+## Solutions - Implementation
+
+### Implement directly within `libelektra-kdb`
+
+Implement all the logic for changetracking directly within `libelektra-kdb`.
+
+### Implement as a seperate plugin
+
+Implement changetracking as a hooks plugin that will be called within `kdbGet` and `kdbSet` accordingly.
+
+The following hooks will be needed:
+
+- `tracking/get`: will be called at the end of `kdbGet`, directly before the result is returned.
+- `tracking/set`: will be called at the beginning of `kdbSet`.
+- `tracking/changeset`: compute the changeset for the requested parent key and return it.
+
+As every hook plugin can define its own contract, in theory all storage forms mentioned in the previous chapter should be possible to implement.
+We could just point the plugin to `backendsData->keys` or the internal cache if we go down that route.
+
+## Solutions - Query
+
+### Provide an API within `libelektra-kdb`
+
+The API should be useable both by plugins and applications utilizing ELektra.
+The API may look something like this:
+
+```c
+bool elektraChangeTrackingIsEnabled (KDB * kdb);
+ChangeTrackingContext * elektraChangeTrackingGetContext (KDB * kdb, Key * parentKey);
+
+KeySet * elektraChangeTrackingGetAddedKeys (ChangeTrackingContext * context);
+KeySet * elektraChangeTrackingGetRemovedKeys (ChangeTrackingContext * context);
+KeySet * elektraChangeTrackingGetModifiedKeys (ChangeTrackingContext * context);
+
+bool elektraChangeTrackingValueChanged (ChangeTrackingContext * context, Key * key);
+bool elektraChangeTrackingMetaChanged (ChangeTrackingContext * context, Key * key);
+
+KeySet * elektraChangeTrackingGetAddedMetaKeys (ChangeTrackingContext * context, Key * key);
+KeySet * elektraChangeTrackingGetRemovedMetaKeys (ChangeTrackingContext * context, Key * key);
+KeySet * elektraChangeTrackingGetModifiedMetaKeys (ChangeTrackingContext * context, Key * key);
+
+Key * elektraChangeTrackingGetOriginalKey (ChangeTrackingContext * context, Key * key);
+const Key * elektraChangeTrackingGetOriginalMetaKey (ChangeTrackingContext * context, Key * key, const char * metaName);
```

Comment:
Please add some context why this would be needed (connect to use case etc.). Isn't three KeySets with all added/removed/modified (meta)keys enough?

Reply:
Good catch! Yes, if `elektraChangeTrackingGetModifiedKeys` / `elektraChangeTrackingGetModifiedMetaKeys` return the *old* keys, then these functions are not required.